client/lib/create-cached-selector/index.js

```diff
@@ -0,0 +1,102 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { isObject, forEach, some, isFunction } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import warn from 'lib/warn';
+
+/*
+ * A map that is Weak with objects but Strong with primitives
+ */
+export class MixedMap {
+	weakMap = new WeakMap();
+	map = new Map();
+
+	constructor( init ) {
+		forEach( init, ( [ key, val ] ) => {
+			this.mapForKey( key ).set( key, val );
+		} );
+	}
+
+	mapForKey = key => ( isObject( key ) ? this.weakMap : this.map );
+
+	clear() {
+		this.weakMap = new WeakMap();
+		this.map.clear();
+		return this;
+	}
+
+	set( k, v ) {
+		this.mapForKey( k ).set( k, v );
+		return this;
+	}
+
+	delete( k ) {
+		this.mapForKey( k ).delete( k );
+		return this;
+	}
+
+	get( k ) {
+		return this.mapForKey( k ).get( k );
+	}
+
+	has( k ) {
+		return this.mapForKey( k ).has( k );
+	}
+}
+
+/**
+ * Returns a selector that caches values.
+ *
+ * @param  {Function} selector      A standard selector for calculating cached result
+ * @param  {Function} getDependents A Function describing the dependent(s) of the selector.
+ *                                    Must return an object which gets used as the first arg of the selector
+ * @return {Function}               Cached selector
+ */
+export default function createCachedSelector( { selector, getDependents } ) {
+	if ( ! isFunction( selector ) || ! isFunction( getDependents ) ) {
+		throw new TypeError(
+			'createCachedSelector: invalid arguments passed, selector and getDependents must both be functions'
+		);
+	}
+
+	const cache = new MixedMap();
+
+	const cachedSelector = function( state, ...args ) {
+		const dependents = getDependents( state, ...args );
+		const sortedDependentsArray = Object.keys( dependents )
+			.sort()
+			.map( key => dependents[ key ] );
+
+		if ( process.env.NODE_ENV !== 'production' ) {
+			if ( some( args, isObject ) ) {
+				warn( 'Do not pass complex objects as arguments to a cachedSelector' );
+			}
+		}
+
+		// create a dependency tree for caching selector results.
+		// this is beneficial over standard memoization techniques so that we can
+		// garbage collect any values that are based on outdated dependents
+		let currCache = cache;
+		forEach( sortedDependentsArray, dependent => {
+			if ( ! currCache.has( dependent ) ) {
+				currCache.set( dependent, new MixedMap() );
```

Comment:
this line may mutate `cache` but it doesn't look like it from here. this is confusing to me and relates to my comment above in `MixedMap`

Reply:
I'd love input on how to do that! I have no idea 🤔 

I like the word `cursor` for what I'm doing with currCache, its more descriptive

Reply:
using `cursor` as the name would help I think.

this is really a kind of insertion on a prefix tree where the dependents are the atoms forming the nodes. things could potentially be more clear simply with the proper messaging around them…

```js
const upsert = ( cache, value ) => {
	if ( ! cache.has( value ) ) {
		cache.set( value, new MixedMap() )
	}

	return cache.get( value );
}

// insert dependents into cache tree
sortedDependentsArray.reduce( upsert, cache )
```

maybe even something along these lines…

Reply:
@samouri did you see this last comment as it pertains to the naming of the `cursor`? (it eliminates it, for example)

Reply:
Now I do! I think its brilliant! I'm just changing the name to `insertIfAbsent` but otherwise keeping the logic intact.